pkg/gardenlet/controller/federatedseed/shootstate/common_sync.go

```diff
@@ -0,0 +1,156 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func createEnqueueOnAddFunc(queue workqueue.RateLimitingInterface) func(extensionObject interface{}) {
+	return func(newObj interface{}) {
+		extensionObject := newObj.(extensionsv1alpha1.Object)
+		if extensionObject.GetExtensionStatus().GetState() == nil {
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func createEnqueueOnUpdateFunc(queue workqueue.RateLimitingInterface) func(newExtensionObject, oldExtensionObject interface{}) {
+	return func(newObj, oldObj interface{}) {
+		var (
+			newExtensionObj = newObj.(extensionsv1alpha1.Object)
+			oldExtensionObj = oldObj.(extensionsv1alpha1.Object)
+		)
+
+		if newExtensionObj.GetExtensionStatus().GetState() == nil {
+			return
+		}
+
+		if oldExtensionObj.GetExtensionStatus().GetState() == nil {
+			enqueue(queue, newObj)
+		}
+
+		if apiequality.Semantic.DeepEqual(newExtensionObj.GetExtensionStatus().GetState().Raw, oldExtensionObj.GetExtensionStatus().GetState().Raw) {
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func enqueue(queue workqueue.RateLimitingInterface, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	queue.Add(key)
+}
+
+func (s *SyncController) createShootStateSyncReconcileFunc(ctx context.Context, kind string, obj runtime.Object) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		obj = obj.DeepCopyObject()
+		err := s.seedClient.Client().Get(ctx, req.NamespacedName, obj)
+		if apierrors.IsNotFound(err) {
+			s.log.Debugf("Skipping ShootState sync because resoruce with kind %s is missing", kind)
+			return reconcile.Result{}, nil
+		}
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		extensionObject, err := apiextensions.Accessor(obj)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		state := extensionObject.GetExtensionStatus().GetState()
+		if state == nil {
+			return reconcile.Result{}, fmt.Errorf("invalid state. Extension state for resource with kind %s in namespace %s was changed to nil before it could be synced to ShootState", kind, req.Namespace)
+		}
+
+		name := extensionObject.GetName()
+		purpose := getPurpose(obj)
+		projectNamespace, shootName := fromNamespaceName(req.Namespace)
+		shootState := &gardencorev1alpha1.ShootState{ObjectMeta: metav1.ObjectMeta{Name: shootName, Namespace: projectNamespace}}
+		if _, err := controllerutil.CreateOrUpdate(ctx, s.k8sGardenClient.Client(), shootState, func() error {
+			return updateShootStateExtensionState(state, shootState, kind, name, purpose)
+		}); err != nil {
+			message := fmt.Sprintf("Shoot %s extension state was NOT succcessfuly synced", kind)
```

Comment:
```suggestion
			message := fmt.Sprintf("Shoot %s extension state was NOT successfully synced", kind)
```

Also, shouldn't we include the error message here to make the event more meaningful?

Reply:
additionally the name and the purpose so that we don't get confused when kind is `Extension` or `ControlPlane`